src/cipher.rs

```diff
@@ -510,296 +509,181 @@ impl rustls_root_cert_store {
-/// `rustls_allow_any_anonymous_or_authenticated_client_verifier` once ready.
-pub struct rustls_allow_any_anonymous_or_authenticated_client_builder {
-    _private: [u8; 0],
-}
-
-impl CastPtr for rustls_allow_any_anonymous_or_authenticated_client_builder {
-    // NOTE: contained value is consumed even on error, so this can contain None. but the caller
-    // still needs to free it
-    type RustType = Option<AllowAnyAnonymousOrAuthenticatedClient>;
-}
+impl BoxCastPtr for rustls_web_pki_client_cert_verifier_builder {}
 
-impl BoxCastPtr for rustls_allow_any_anonymous_or_authenticated_client_builder {}
-
-impl rustls_allow_any_anonymous_or_authenticated_client_builder {
-    /// Create a new allow any anonymous or authenticated client certificate verifier builder
-    /// using the root store.
+impl rustls_web_pki_client_cert_verifier_builder {
+    /// Create a `rustls_web_pki_client_cert_verifier_builder`. Caller owns the memory and must
+    /// eventually call `rustls_web_pki_client_cert_verifier_builder_build`, then free the
+    /// resulting `rustls_client_cert_verifier`.
+    ///
+    /// Without further modification the builder will produce a client certificate verifier that
+    /// will require a client present a client certificate that chains to one of the trust anchors
+    /// in the provided `rustls_root_cert_store`. The root cert store must not be empty.
     ///
-    /// This copies the contents of the rustls_root_cert_store. It does not take
-    /// ownership of the pointed-to memory.
+    /// Revocation checking will not be performed unless
+    /// `rustls_web_pki_client_cert_verifier_builder_add_crl` is used to add certificate revocation
+    /// lists (CRLs) to the builder.
     ///
-    /// This object can then be used to load any CRLs.
+    /// Anonymous unauthenticated clients will not be permitted unless
+    /// `rustls_web_pki_client_cert_verifier_builder_allow_unauthenticated` is used.
     ///
-    /// Once that is complete, convert it into a real
-    /// `rustls_allow_any_anonymous_or_authenticated_client_verifier`
-    /// by calling `rustls_allow_any_anonymous_or_authenticated_client_verifier_new()`.
+    /// This copies the contents of the `rustls_root_cert_store`. It does not take
+    /// ownership of the pointed-to data.
     #[no_mangle]
-    pub extern "C" fn rustls_client_cert_verifier_optional_builder_new(
+    pub extern "C" fn rustls_web_pki_client_cert_verifier_builder_new(
         store: *const rustls_root_cert_store,
-    ) -> *mut rustls_allow_any_anonymous_or_authenticated_client_builder {
+    ) -> *mut rustls_web_pki_client_cert_verifier_builder {
         ffi_panic_boundary! {
-            let store: &RootCertStore = try_ref_from_ptr!(store);
-            let client_cert_verifier = Some(AllowAnyAnonymousOrAuthenticatedClient::new(store.clone()));
-            BoxCastPtr::to_mut_ptr(client_cert_verifier)
+            let store: Arc<RootCertStore> = try_arc_from_ptr!(store);
```

Comment:
I'm seeing a crash running the client server integration tests that I think is happening on this line. Specifically it's happening with the first test that sets an `AUTH_CERT` for the server to enforce mandatory client authentication. 

```
$> make clean && make && AUTH_CERT=testdata/minica.pem gdb --args ./target/server testdata/localhost/cert.pem testdata/localhost/key.pem
<snipped>
(gdb) run
Starting program: /home/daniel/Code/Rust/rustls-ffi/target/server testdata/localhost/cert.pem testdata/localhost/key.pem
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libthread_db.so.1".

Program received signal SIGILL, Illegal instruction.
0x00000000004d3fc9 in std::panicking::try::h608246b5500614b5 ()
(gdb) bt
#0  0x00000000004d3fc9 in std::panicking::try::h608246b5500614b5 ()
#1  0x000000000040a35d in rustls_web_pki_client_cert_verifier_builder_new ()
#2  0x00000000004062ef in main (argc=<optimized out>, argv=<optimized out>) at tests/server.c:293
(gdb) q
```

Will debug further.

Reply:
<details>
<summary>better backtrace:</summary>

```
$> make clean && make PROFILE=debug && AUTH_CERT=testdata/minica.pem gdb --args ./target/server testdata/localhost/cert.pem testdata/localhost/key.pem
<snipped>
(gdb) run
Starting program: /home/daniel/Code/Rust/rustls-ffi/target/server testdata/localhost/cert.pem testdata/localhost/key.pem
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libthread_db.so.1".

Program received signal SIGILL, Illegal instruction.
alloc::sync::{impl#22}::clone<rustls::webpki::anchors::RootCertStore> (self=0x7fffffff4d50) at /build/rustc-1.72.0-src/library/alloc/src/sync.rs:1530
1530	/build/rustc-1.72.0-src/library/alloc/src/sync.rs: No such file or directory.
(gdb) bt
#0  alloc::sync::{impl#22}::clone<rustls::webpki::anchors::RootCertStore> (self=0x7fffffff4d50) at /build/rustc-1.72.0-src/library/alloc/src/sync.rs:1530
#1  0x0000000000406c74 in rustls_ffi::ArcCastPtr::to_arc<rustls_ffi::cipher::rustls_root_cert_store> (ptr=0x6be410) at src/lib.rs:377
#2  0x000000000040996b in rustls_ffi::try_arc_from<rustls_ffi::cipher::rustls_root_cert_store, rustls::webpki::anchors::RootCertStore> (from=0x6be410) at src/lib.rs:456
#3  0x0000000000407541 in rustls_ffi::cipher::{impl#16}::rustls_web_pki_client_cert_verifier_builder_new::{closure#0} () at src/cipher.rs:584
#4  0x0000000000414d0e in std::panicking::try::do_call<rustls_ffi::cipher::{impl#16}::rustls_web_pki_client_cert_verifier_builder_new::{closure_env#0}, *mut rustls_ffi::cipher::rustls_web_pki_client_cert_verifier_builder> (data=0x7fffffff4ee0)
    at /build/rustc-1.72.0-src/library/std/src/panicking.rs:500
#5  0x000000000041609b in __rust_try ()
#6  0x0000000000413e13 in std::panicking::try<*mut rustls_ffi::cipher::rustls_web_pki_client_cert_verifier_builder, rustls_ffi::cipher::{impl#16}::rustls_web_pki_client_cert_verifier_builder_new::{closure_env#0}> (f=...) at /build/rustc-1.72.0-src/library/std/src/panicking.rs:464
#7  0x0000000000411c97 in std::panic::catch_unwind<rustls_ffi::cipher::{impl#16}::rustls_web_pki_client_cert_verifier_builder_new::{closure_env#0}, *mut rustls_ffi::cipher::rustls_web_pki_client_cert_verifier_builder> (f=...) at /build/rustc-1.72.0-src/library/std/src/panic.rs:142
#8  0x0000000000408ad3 in rustls_ffi::cipher::rustls_web_pki_client_cert_verifier_builder::rustls_web_pki_client_cert_verifier_builder_new (store=0x6be410) at src/cipher.rs:583
#9  0x00000000004046bf in main (argc=<optimized out>, argv=<optimized out>) at tests/server.c:293
```
</details>

Reply:
https://github.com/rust-lang/rust/blob/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/sync.rs#L1500-L1535

Reply:
It's certainly the case that I've messed up the `Arc<RootCertStore>` handling introduced in this branch, but I'm yet sure what changed between this most recent force push and https://github.com/rustls/rustls-ffi/commit/07532681e838e6e1d8ecd9890273e14eeacb22b2 when it at least appeared to work :thinking: 

The test seems to work if I introduce a clumsy diff like this, but I'm not confident it's a well reasoned change:
```diff
diff --git a/src/cipher.rs b/src/cipher.rs
index b5e9eb4..b62bdeb 100644
--- a/src/cipher.rs
+++ b/src/cipher.rs
@@ -578,12 +578,13 @@ impl rustls_web_pki_client_cert_verifier_builder {
     /// ownership of the pointed-to data.
     #[no_mangle]
     pub extern "C" fn rustls_web_pki_client_cert_verifier_builder_new(
-        store: *const rustls_root_cert_store,
+        store: *mut rustls_root_cert_store,
     ) -> *mut rustls_web_pki_client_cert_verifier_builder {
         ffi_panic_boundary! {
-            let store: Arc<RootCertStore> = try_arc_from_ptr!(store);
+            //let store: Arc<RootCertStore> = try_arc_from_ptr!(store);
+            let store: &mut RootCertStore = try_mut_from_ptr!(store);
             let builder = ClientCertVerifierBuilder {
-                roots: store,
+                roots: Arc::new(store.clone()),
                 crls: Vec::default(),
                 allow_anonymous: false,
             };
```

Reply:
You're on the right track. I wasn't able to get this completely working before having to sign off, but two key things:

 - try_arc_from_ptr is what you want for any time you want to do clones of an arc. See docs on [ArcCastPtr::to_arc](https://github.com/rustls/rustls-ffi/blob/6c24d4829e1d596ce03e6097145e023be29acf78/src/lib.rs#L356-L371), which it calls under the hood. It does some refcounting fixups.
 - Only one of BoxCastPtr or ArcCastPtr should be implemented for a type; the fact that BoxCastPtr was implemented prevented the type system from catching the problem here.

The below doesn't compile yet, but should indicate a good direction for progress yet; happy to help more tomorrow.

```
diff --git a/src/cipher.rs b/src/cipher.rs
index b5e9eb4a..1325f8d3 100644
--- a/src/cipher.rs
+++ b/src/cipher.rs
@@ -443,8 +443,6 @@ impl CastPtr for rustls_root_cert_store {
     type RustType = RootCertStore;
 }
 
-impl BoxCastPtr for rustls_root_cert_store {}
-
 impl ArcCastPtr for rustls_root_cert_store {}
 
 impl rustls_root_cert_store {
@@ -453,10 +451,10 @@ impl rustls_root_cert_store {
     /// Caller must add root certificates with rustls_root_cert_store_add_pem.
     /// <https://docs.rs/rustls/latest/rustls/struct.RootCertStore.html#method.empty>
     #[no_mangle]
-    pub extern "C" fn rustls_root_cert_store_new() -> *mut rustls_root_cert_store {
+    pub extern "C" fn rustls_root_cert_store_new() -> *const rustls_root_cert_store {
         ffi_panic_boundary! {
             let store = rustls::RootCertStore::empty();
-            BoxCastPtr::to_mut_ptr(store)
+            ArcCastPtr::to_const_ptr(&store)
         }
     }
 
@@ -503,7 +501,7 @@ impl rustls_root_cert_store {
     #[no_mangle]
     pub extern "C" fn rustls_root_cert_store_free(store: *mut rustls_root_cert_store) {
         ffi_panic_boundary! {
-            let store = try_box_from_ptr!(store);
+            let store = try_arc_from_ptr!(store);
             drop(store)
         }
     }
@@ -519,7 +517,7 @@ impl CastPtr for rustls_client_cert_verifier {
     type RustType = Arc<dyn ClientCertVerifier>;
 }
 
-impl BoxCastPtr for rustls_client_cert_verifier {}
+impl ArcCastPtr for rustls_client_cert_verifier {}
 
 impl rustls_client_cert_verifier {
     /// Free a `rustls_client_cert_verifier` previously returned from
@@ -583,7 +581,7 @@ impl rustls_web_pki_client_cert_verifier_builder {
         ffi_panic_boundary! {
             let store: Arc<RootCertStore> = try_arc_from_ptr!(store);
             let builder = ClientCertVerifierBuilder {
-                roots: store,
+                roots: store.clone(),
                 crls: Vec::default(),
                 allow_anonymous: false,
             };
diff --git a/src/lib.rs b/src/lib.rs
index c002925f..9c0b5f56 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -329,7 +329,8 @@ where
 }
 
 // An implementation of BoxCastPtr means that when we give C code a pointer to the relevant type,
-// it is actually a Box.
+// it is actually a Box. At most one of BoxCastPtr or ArcCastPtr should be implemented for a given
+// type.
 pub(crate) trait BoxCastPtr: CastPtr + Sized {
     fn to_box(ptr: *mut Self) -> Option<Box<Self::RustType>> {
         if ptr.is_null() {
diff --git a/src/server.rs b/src/server.rs
index 4a46319f..94cd271b 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -166,7 +166,7 @@ impl rustls_server_config_builder {
     ) {
         ffi_panic_boundary! {
             let builder: &mut ServerConfigBuilder = try_mut_from_ptr!(builder);
-            let verifier: Arc<dyn ClientCertVerifier> = try_ref_from_ptr!(verifier).clone();
+            let verifier: Arc<dyn ClientCertVerifier> = try_arc_from_ptr!(verifier);
             builder.verifier = verifier;
         }
     }
```

Reply:
> You're on the right track. I wasn't able to get this completely working before having to sign off, but two key things:

Thanks Jsha!

> Only one of BoxCastPtr or ArcCastPtr should be implemented for a type; the fact that BoxCastPtr was implemented prevented the type system from catching the problem here.

Oh interesting! I had done the same in my earlier revision that seemed to work. I suspect that it was just more subtly broken than the current branch :grimacing: 

I will pick up your diff today and see where I can get. Appreciate the guidance :bow:

Reply:
I have a few questions after playing with the diff some more. Sorry if some of this is obvious. I'm still trying to get the right mental model in place and find the dance between Rust types, the helper macros, and the C semantics tricky.

I've pushed some revisions and I think it's closer: the rust code builds. The `server.c` code fails from discarding the `const` qualifier on the `rustls_root_cert_store` pointer based on the "Root Cert store `ArcCastPtr`" problem discussed below.

### Root Cert store `ArcCastPtr`

```diff
-impl BoxCastPtr for rustls_root_cert_store {}
-
 impl ArcCastPtr for rustls_root_cert_store {}
 ```
 
The `ArcCastPtr` change on `rustls_root_cert_store` implies changing `rustls_root_cert_store_new` and `rustls_root_cert_store_free` to work with `*const rustls_root_cert_store`, but that's not compatible with the needs of `rustls_root_cert_store_add_pem` where we want a `*mut rustls_root_cert_store` to use with `try_mut_from_ptr!`, right? I'm not sure how to reconcile this now that I understand `BoxCastPtr` and `ArcCastPtr` are mutually exclusive.

### Explicit `Arc<RootCertStore>` clone

```diff
 impl rustls_client_cert_verifier {
     /// Free a `rustls_client_cert_verifier` previously returned from
@@ -583,7 +581,7 @@ impl rustls_web_pki_client_cert_verifier_builder {
         ffi_panic_boundary! {
             let store: Arc<RootCertStore> = try_arc_from_ptr!(store);
             let builder = ClientCertVerifierBuilder {
-                roots: store,
+                roots: store.clone(),
                 crls: Vec::default(),
                 allow_anonymous: false,
             };
```

Clippy finds the `clone` here unnecessary. Is it missing something?

### `Arc<Arc<dyn ClientCertVerifier>>`

```diff
diff --git a/src/server.rs b/src/server.rs
index 4a46319f..94cd271b 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -166,7 +166,7 @@ impl rustls_server_config_builder {
     ) {
         ffi_panic_boundary! {
             let builder: &mut ServerConfigBuilder = try_mut_from_ptr!(builder);
-            let verifier: Arc<dyn ClientCertVerifier> = try_ref_from_ptr!(verifier).clone();
+            let verifier: Arc<dyn ClientCertVerifier> = try_arc_from_ptr!(verifier);
             builder.verifier = verifier;
         }
     }
```

Using `try_arc_from_ptr!` here gives us an `Arc<Arc<dyn ClientCertVerifier>>`, whereas `try_ref_from_ptr!` returns the `Arc<dyn ClientCertVerifier>` the `builder.base.with_client_cert_verifier` fn that the `verifier` field is ultimately used for requires. I thought perhaps changing the `rustls_client_cert_verifier`'s `RustType` to be `dyn ClientCertVerifier` instead of `Arc<dyn ClientCertVerifier>` would be the way to use `try_arc_from_ptr!` here, but then the size of the type isn't known at compile time. Using a reference introduces complicates related to the lifetime of the reference, so it doesn't seem like the right approach either. I think this is the first instance of a `CastPtr` for a trait instead of a struct so there isn't existinv code to crib from. 

### Out parameter for an `ArcCastPtr`

In [earlier discussions](https://github.com/rustls/rustls-ffi/pull/341#issuecomment-1673836497) we landed on using an out parameter pattern for the client cert verifier's build operation, since its fallible and needs to use the traditional return for a `rustls_result`.

The existing instances of this pattern are using types that are `BoxCastPtr`, but the new builder is now only `ArcCastPtr` and so can't use the `BoxCastPtr::set_mut_ptr` helper. Is the correct approach to set the out parameter type to be `*mut *const rustls_client_cert_verifier` (as opposed to `*mut *mut rustls_client_cert_verifier` ala the other `BoxCastPtr` instances) and then use an unsafe block similar to the`set_mut_ptr` helper? Something like:

```rust
 unsafe {
      *verifier_out = ArcCastPtr::to_const_ptr(verifier)
  }
```

Reply:
> that's not compatible with the needs of rustls_root_cert_store_add_pem where we want a *mut rustls_root_cert_store to use with try_mut_from_ptr!, right?

Yep, that's right; so there's some more API design work to be done here. The basic idea of BoxCastPtr / ArcCastPtr is "any time we give C a pointer of type X, it really corresponds to a Box<T> under the hood" or "... really corresponds to an Arc<T> under the hood".

If you look at it from the Rust side, you have the same problem: If you've got an `Arc<T>` you can't modify it. Instead, you do all the modification you need ahead of time, and then wrap the resulting T in an `Arc`. That's why our `Arc<T>` types are represented on the C side with a `*const`, to clarify that "can't modify" situation.

So I think what we need here is another builder pattern: create a root_cert_store_builder, add certs to it, then "build" it by wrapping it in an Arc.

(more soon)

Reply:
> Clippy finds the clone here unnecessary. Is it missing something?

Clippy is correct here. Instead of cloning the Arc we got out of `ArcCastPtr::to_arc`, we can simply hand over ownership of it to the ClientCertVerifierBuilder. By doing that, the Arc will no longer be dropped at the end of the function (because it was moved instead), so the refcount won't be decremented. It will remain at least 2 (1 for the C side, 1 for this particular reference on the Rust side).

> Out parameter for an ArcCastPtr

Your approach here is correct. I think to refine on it, we should add a `set_mut_ptr` method on the ArcCastPtr trait and use that.

> Arc<Arc<dyn ClientCertVerifier>>

Looking at this in more detail now.

Reply:
> whereas try_ref_from_ptr! returns the Arc<dyn ClientCertVerifier>

Actually, `try_ref_from_ptr!(verifier:  *const rustls_client_cert_verifier)` is returning an `&Arc<dyn ClientCertVerifier>`, which is correct and what I'd expect. We can then call `clone()` on that reference (since [`Arc::clone`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.clone) takes `&self`).

Perhaps the other site where I proposed `try_arc_from_ptr!` would be equally well served by `try_ref_from_ptr!` plus `clone()`; we might actually be able to do away with `try_arc_from_ptr!` entirely, which would be a nice clarification.

Reply:
I think I've rolled in all of this feedback/learning now :rocket: The commit history has gotten a little bit messy but otherwise I think it's in a workable state now :sweat_smile: 

> Clippy is correct here. Instead of cloning the Arc we got out of ArcCastPtr::to_arc, we can simply hand over ownership of it to the ClientCertVerifierBuilder

This ended up falling away after switching to `try_ref_from_ptr!` in more places.

> Actually, try_ref_from_ptr!(verifier:  *const rustls_client_cert_verifier) is returning an &Arc<dyn ClientCertVerifier>, which is correct and what I'd expect. We can then call clone() on that reference

Thanks, that's what I landed on too. I should have been more precise in my description :+1: 

> Your approach here is correct. I think to refine on it, we should add a set_mut_ptr method on the ArcCastPtr trait and use that.

Done :ballot_box_with_check: 

> we might actually be able to do away with try_arc_from_ptr! entirely, which would be a nice clarification.

I think I've managed to achieve this :crossed_fingers: